### PR TITLE
Replace superuser role for API keys

### PR DESCRIPTION
**Pull request by ywangd**
This PR replaces superuser role of an API key with the new limited
superuser role to prevent write access to system indices.

The replacement should only happen to the builtin superuser role. 
If there is a user-created role that has the exact same definition as
the superuser role, it will not be replaced because we consider users
are explicitly opt-in for ALL access in this scenario. 

Since API key captures (snapshots) user roles when creating the key,
it is not always straightforward to decide whether the captured role descriptor
is indeed the builtin superuser, e.g. before #82049, the key role descriptors can 
be created exactly like the builtin superuer (including the `_reserved` metadata 
field). The solution is to only replace role descriptors from the limited-by roles because:
* Currently the limited-by roles are either sourced from the role providers using "names"
   or they are empty (derived keys)
* If we decide to expand support for derived keys in future, because of #82049, new keys
  will not have be able to have role descriptors with `_reserved` metadata field. For old 
  keys, we can add additional checks to drop any invalid metadata field (or throw errors).
  In short, we should be able to prevent a custom role looks exactly like the builtin superuser
  role.

Relates: #81400

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java [hunk 1]
```diff
@@ -133,12 +133,12 @@ public Type getType() {
...
             apiKeyId,
             limitedByRoleDescriptorsBytes,
-            "apikey_limited_role"
+            true
         );
         if (isEmptyRoleDescriptorsBytes(roleDescriptorsBytes)) {
             return List.of(limitedByRoleReference);
         }
-        return List.of(new RoleReference.ApiKeyRoleReference(apiKeyId, roleDescriptorsBytes, "apikey_role"), limitedByRoleReference);
+        return List.of(new RoleReference.ApiKeyRoleReference(apiKeyId, roleDescriptorsBytes, false), limitedByRoleReference);
```
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java:141]: It's probably a _nit_, but I think an enum would be better than a `boolean`. Something like:
```
enum ApiKeyRoleType {
  ASSIGNED_ROLES, LIMITED_BY_ROLES
}
```
(I'm happy with different names for them enum class, or members)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java [hunk 2]
```diff
@@ -568,7 +575,40 @@ void loadApiKeyAndValidateCredentials(
...
+        MetadataUtils.DEFAULT_RESERVED_METADATA,
+        Collections.emptyMap()
+    );
+
+    // This method should only be called to replace the superuser role descriptor for the limited-by roles of an API Key.
+    // We do not replace assigned roles because they are created explicitly by users.
+    // Before #82049, it is possible to specify a role descriptor for API keys that is identical to the builtin superuser role
+    // (including the _reserved metadata field).
+    private List<RoleDescriptor> maybeReplaceSuperuserRoleDescriptor(String apiKeyId, List<RoleDescriptor> roleDescriptors) {
+        // Scan through all the roles because superuser can be one of the roles that a user has. Unlikely building the Role object,
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:600]:
```suggestion
        // Scan through all the roles because superuser can be one of the roles that a user has. Unlike building the Role object,
```

**Review by tvernum — COMMENTED** [review]
I'm happy with the approach.
I haven't reviewed 100% of the changes, but I can do a final pass in the morning.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java (hunk 1 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java]: I added the enum and also use the enum toString for the source of the roleKey to avoid having extra string constants. It might also become useful if later we decide to support layered limited-by. I think it's an overall improvement. Thanks for the suggestion.

**Comment by ywangd** [comment]
@elasticmachine update branch

**Review by tvernum — APPROVED** [review]
LGTM

**Comment by ywangd** [comment]
CI Failure is not related and not reproducible. I raised https://github.com/elastic/elasticsearch/issues/82647

**Comment by ywangd** [comment]
@elasticmachine run elasticsearch-ci/part-1
